### PR TITLE
Make it possible to export container dependant variables

### PR DESCRIPTION
**Pull request by hexa00**
This patch introduces the initialize function inside a module, this
function is called with the root container of the application at
application startup.

This enables the logger module to export a logger object fetched via
container.get<ILogger>(ILogger) such that this new logger object can be
exported to code that does not use injection.

Signed-off-by: Antoine Tremblay <antoine.tremblay@ericsson.com>

### CODE REVIEW

**Comment by hexa00** [comment]
See: #517

**Comment by akosyakov** [comment]
I don't think we should generalize it, a case with the root logger is special. It would be nice to keep the module interface simple. Also the first thing to access from the root container must be an application, otherwise the root container is not considered complete, meaning that other extensions could override some services require for logging after loading the loader module. 

Is there a reason not to do in `onActivation` of an application?
Like here: https://github.com/inversify/InversifyJS/blob/master/wiki/activation_handler.md#activation-handler

**Comment by svenefftinge** [comment]
I agree with Anton, I think it should just be a side-effect in the container's load.

**Comment by hexa00** [comment]
Ha yes indeed onActivation is what I need, thanks!

**Comment by hexa00** [comment]
The latest change is WIP , I don't understand how to make this global var work, ideas ?
Note the terminal-process changes are just there for testing
the logger is undefined when trying to log the terminal creation
You can test by starting a term

**Comment by svenefftinge** [comment]
`onActivation` is called on creation of a logger. If it is not demanded through @inject it is not created. I think @akosyakov meant to use `onActivation` of `FrontendApplication` resp. `BackendApplication` bindings.

**Comment by hexa00** [comment]
@svenefftinge but if I do that the application will need to depend on the logger it works for core but it would not work for another package.

Also independenly of that do you have an idea for the gobal export ?

**Comment by hexa00** [comment]
Note we will want do that with the localize package also...

**Comment by svenefftinge** [comment]
I think you would need to rebind the FrontendApplication within the logger container module and add an onActivation hook. But I'm not so happy with that, too. Could you outline why we need that for the localization? Maybe with two cases, it is worth to introduce a callback as you did in the first PR. :)

**Comment by hexa00** [comment]
In the case of the localization we want to make it easy to localize a string anywhere in theia.

So you could call localize("mystring"); directly from anywhere without injecting the MessageTranslation class (localize is exported from that class).

But we need the MessageTranslation class in the inversify context so that it can access the user preferences to select the proper language.

So we need to bridge the gap between the inversify context and static context same as with the logger.

**Comment by hexa00** [comment]
Also localization is it's own package that would be upstream of core.

**Comment by hexa00** [comment]
Actually thinking about this now localize can't depend on preferences since preferences depend on too much stuff... I'm not sure how to structure this properly yet it seems :(

**Comment by hexa00** [comment]
Actually for localization I think we could have a mock preferences class until we get the real preferences and than rebind those.

So it would properly once you get the whole application but otherwise use a failover.

**Comment by svenefftinge** [comment]
I thought we were doing this not for convenience, but because there we need to log in a static context?

**Comment by svenefftinge** [comment]
adding injection for logger and localize should be the default. For localize we should try to avoid doing this global scope hack.

**Comment by hexa00** [comment]
For localize it would mean basically every class would have a localize injected given that you would want to localize log messages too possibly.

That's what I'm trying to avoid.

**Comment by hexa00** [comment]
Note the logger is pretty much as oblivious

**Comment by svenefftinge** [comment]
I think we should only localize user-facing messages. logs are for developers.

**Comment by hexa00** [comment]
Maybe that restrains the places where it's used enought...

**Comment by akosyakov** [comment]
Why should it be global, is not possible to use the module import?

**Comment by akosyakov** [comment]
BTW what about an idea to rebind console functions instead of exposing our logger?

**Comment by hexa00** [comment]
For localize it has to be gobal to do like this:
https://github.com/lmcbout/theia/blob/2b05667dbcc565c42f2f71d583fe4c4173eb405f/packages/workspace/src/browser/workspace-frontend-contribution.ts

See:
```
export namespace WorkspaceCommands {
    export const OPEN: Command = {
        id: 'workspace:open',
        label: localize("Open", "Open...")
    };
}
```

**Comment by hexa00** [comment]
For the logger there is less reason to be global at the moment since I don't have a static context to call it from like I had before.

But I wanted to do the logger as an example for localize to follow and provide a general way to do this when need be for anything.

**Comment by svenefftinge** [comment]
Ok, do we need localize in the container, then?

**Comment by svenefftinge** [comment]
I guess you want to use server-side preferences for the locale?

**Comment by hexa00** [comment]
>Ok, do we need localize in the container, then?

Not sure what you mean ?

>I guess you want to use server-side preferences for the locale?
Yes that's the plan and this is another issue since core has strings to localize but if localize depends on preferences and preferences depend on core/filesystem/workspace it's a problem... but let's tackle that later.

For now I'm trying to figureout out how an extension can contribute a global to be able to call something in an inversify context from static context.

**Comment by akosyakov** [comment]
> For now I'm trying to figureout out how an extension can contribute a global to be able to call something in an inversify context from static context.

Why should it be global? Globals are generally discouraged to avoid namespace conflicts. Why not use the module import as you already do for `localize`, e.g.:

```ts
import { logger } from '@theia/core/common/logger';

logger.log('lalala');
```

**Comment by akosyakov** [comment]
> I guess you want to use server-side preferences for the locale?
>> Yes that's the plan and this is another issue since core has strings to localize but if localize depends on preferences and preferences depend on core/filesystem/workspace it's a problem... but let's tackle that later.

Don't you need to repack everything on locale change?

**Comment by hexa00** [comment]
>Why should it be global? Globals are generally discouraged to avoid namespace conflicts. Why not use the module import as you already do for localize, e.g.:

I think there's some confusion about what I mean by global.. yes like your example I mean a global as in:
```
common/logger.ts
export let logger;
```
So more like a module global..

The important thing is that the module can set this varible onActivation like I do in this PR: https://github.com/theia-ide/theia/pull/562/files

But that doesn't work for me at the moment... somehow I end up with undefined in my test in the TerminalProcess when I try to import logger and use it... ideas on that?

**Comment by hexa00** [comment]
>Don't you need to repack everything on locale change?

No you don't this can be changed dynamically. All the languages are shipped with theia and can be loaded on demand.

**Comment by akosyakov** [comment]
`onActivation` is only triggered when someone requests `Logger`, probably in your case nobody requests it

**Comment by hexa00** [comment]
Yes it is requested I can set a breakpoint into setRootLogger and that is beeing called.

**Comment by hexa00** [comment]
But yes that's a problem in general...

**Comment by akosyakov** [comment]
or it is requested too late

**Comment by hexa00** [comment]
Well it should be requested on the container.get(Application) so it's not too late for the terminal.
Also in the debugger I do see setRootLogger be called before the logger usage in terminal process.

**Comment by hexa00** [comment]
Note that if i do this and export logger directly from logger-backend-module it works fine.
But the indirection with commong/logger.ts is causing the issue.

File: packages/process/src/node/terminal-process.ts [hunk 1]
```diff
@@ -47,11 +47,10 @@ export class TerminalProcess extends Process {
 
     constructor(
         @inject(TerminalProcessOptions) options: TerminalProcessOptions,
-        @inject(ProcessManager) processManager: ProcessManager,
-        @inject(ILogger) logger: ILogger) {
+        @inject(ProcessManager) processManager: ProcessManager) {
         super(processManager, logger);
 
-        this.logger.debug(`Starting terminal process: ${options.command},`
+        logger.info(`Starting terminal process: ${options.command},`
```
Comment (akosyakov) [inline packages/process/src/node/terminal-process.ts]: We should use injection as much as possible to keep things simple. I don't think that accessing stuff outside of inversify context should be encouraged.

Comment (hexa00) [reply packages/process/src/node/terminal-process.ts]: This is juste meant as a proof of concept test it's not proposed to be merged

**Comment by akosyakov** [comment]
try to use Promise and resolve it in `onActivation`

**Comment by akosyakov** [comment]
`import logger from 'common/logger.ts` basically create a variable, it is probably the issue

**Comment by hexa00** [comment]
That would make code like:
logger.then(logger => logger.info('....'))
same with localize

I think that would be quite ugly...

So `import logger from 'common/logger.ts' creates a variable you mean that this logger variable is independant from the exported one ?

**Comment by akosyakov** [comment]
I am not sure, you better look at the generated code and debug the import.

**Comment by hexa00** [comment]
According to: http://2ality.com/2015/07/es6-module-exports.html

With CommonJS the exports is done by value rather than by reference, hence the issue I have.

Seems this is not the case with ES6 modules however.

**Comment by hexa00** [comment]
Then again reading https://github.com/Microsoft/TypeScript/issues/6751

If I use require like so:

```
const log = require('@theia/core/lib/common/logger');
log.logger.info(`Starting terminal process: ${options.command},`
```

this works fine.

**Comment by akosyakov** [comment]
and with such syntax?
```ts
import log = require('@theia/core/lib/common/logger');
```

**Comment by akosyakov** [comment]
or `import * as log from '@theia/core/lib/common/logger'`

**Comment by hexa00** [comment]
Actually the problem was that doing
```typescript
import { logger } from '@theia/core/lib/common';
```
Gives:
```javascript
var common_1 = require("@theia/core/lib/common");
common_1.logger.info("Starting terminal process: " + options.command + ","
```
And here logger is undefined

However doing:
```typescript
import { logger } from '@theia/core/lib/common/logger';
```
Gives:

```javascript
var logger_1 = require("@theia/core/lib/common/logger");
logger_1.logger.info("Starting terminal process: " + options.command + ","
```
And this is OK.

So the reexporting in the index is the issue, not sure why ... ideas?

**Comment by akosyakov** [comment]
> With CommonJS the exports is done by value rather than by reference, hence the issue I have.

because of it again? basically, reexport creates a new module and reassign properties from reexported modules, it happens only once

**Comment by akosyakov** [comment]
What if you just reexport functions, like `info`, `error` and so on? e.g.

```ts
import logger from '@theia/core/lib/common/logger';
logger.info('blabla');
```

or 

```ts
import { info } from '@theia/core';
info('blabla');
```

What with the idea to rebind `console.log` to `logger.info`?

**Comment by hexa00** [comment]
So looking at index.js:

```javascript
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./types"));
`
```

So yes this makes an new export object so here it's taken by value and can't be changed.

So I'm just going to directly import it for this case and use:

```typescript
import { logger } from '@theia/core/lib/common/logger';
logger.info('foo');
```

**Comment by hexa00** [comment]
@svenefftinge 
> think you would need to rebind the FrontendApplication within the logger container module and add an onActivation hook. But I'm not so happy with that, too. Could you outline why we need that for the localization? Maybe with two cases, it is worth to introduce a callback as you did in the first PR. :)

Are you ok with adding this, given the issue with the localization I exposed ?

**Comment by svenefftinge** [comment]
It somehow feels like we are going down the wrong route here, but I don't see alternatives :-|
For localization, I am not sure if we really need to do that with backend preferences. Maybe local storage would work well, too.

**Comment by svenefftinge** [comment]
But if you want to move on with the logger, I'm fine with adding the rebind for now.

**Comment by hexa00** [comment]
>It somehow feels like we are going down the wrong route here, but I don't see alternatives :-|

I agree, however at lest the initialze thing will be quite self contained and generic.

>For localization, I am not sure if we really need to do that with backend preferences. Maybe local storage would work well, too.

Yes I tought about that it's just that, I wonder if preferences could have a local storage backend ? so we could use the same API if the filesystem is not present? 

>But if you want to move on with the logger, I'm fine with adding the rebind for now.
Thanks I'll repost the original PR

**Comment by hexa00** [comment]
@svenefftinge back to to originial proposal... is it OK ?

**Comment by svenefftinge** [comment]
It is ok, although if we go with the generic callback variant I would prefer using DI and our `bindContributionProvider`. I.e. define an interface like `Initializer` and register it using `bindContributionProvider`. Then, the first thing you do with a container is to obtain all instances and call their `initialize` method.

**Comment by hexa00** [comment]
That sounds like a good idea, I'll give it a try.

**Comment by hexa00** [comment]
@svenefftinge reworked that with bindContributionProvider... looks better ?

**Comment by akosyakov** [comment]
Could we reuse `BackendApplicationContribution.configure` for it? And add `FrontendApplicationContribution.configure` to be symetric?

**Comment by hexa00** [comment]
@akosyakov no since we can't be certain of the order of the contribution... if another contribution is loaded before and expects things to be initialzed it will fail :(

Unless there's a way to put some priority there.. but I'd like to avoid depending on the order the modules are loaded.

**Comment by akosyakov** [comment]
> no since we can't be certain of the order of the contribution... if another contribution is loaded before and expects things to be initialzed it will fail :(

In which cases? Do you mean other contributions want to use logger? I thought they should inject an instance of `Logger`?

Or add `initialize` callback to `BackendApplicationContribution` and `FrontendApplicationContribution` which is called before `configure`? I thought these interfaces were to provide application lifecycle callbacks.

**Comment by hexa00** [comment]
> In which cases? Do you mean other contributions want to use logger? I thought they should inject an instance of Logger?

Yes but if we give the option that they don't have to inject this, it has to work in all cases.

True I could add initialize to the `Backend/FrontendApplicationContribution` I wonder why I did not think of this before now hehe.

I'll refactor as such.

**Comment by hexa00** [comment]
@akosyakov OK with recent changes ?

File: packages/core/src/node/backend-application.ts [hunk 2]
```diff
@@ -37,6 +38,13 @@ export class BackendApplication {
                 }
             }
         });
+
+        for (const contribution of this.contributionsProvider.getContributions()) {
+            if (contribution.initialize) {
+                contribution.initialize();
```
Comment (akosyakov) [inline packages/core/src/node/backend-application.ts]: could be done separately, but we should also wrap calls in try-catch in the backend

Comment (hexa00) [reply packages/core/src/node/backend-application.ts]: Right I'll make a PR for this now

Comment (hexa00) [reply packages/core/src/node/backend-application.ts]: I'm adding this to this PR after all.. since it's quite trivial

**Pull request merged**